Overmap shuttle refactor

comma opened the pull request:
Makes a proper shuttle subtype for overmap, moves whole bunch of shuttle logic into it from shuttle console.
Also makes it possible to define shuttles yourself instead of letting shuttle consoles generate one.
Does away with concept of 'home' area, overmap shuttles just care about starting area and destination area, no areas are given preferential treatment.
Fixes bunch of wonkiness previous version of spahgetti caused, like consoles holding whole lot of shuttle-used vars in them without syncing, or destination choosing weirdness caused by having special 'home' areas.

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

comma committed:
Overmap shuttle refactor

Makes a proper shuttle subtype for overmap, moves whole bunch of shuttle logic into it from shuttle console.
Does away with concept of 'home' area, overmap shuttles just care about starting area and destination area, no areas are given preferential treatment.
Fixes bunch of wonkiness previous version of spahgetti caused, like consoles holding whole lot of shuttle-used vars in them without syncing, or destination choosing weirdness caused by having special 'home' areas.

PsiOmegaDelta commented on code/modules/overmap/ships/computers/shuttle.dm:
```diff
@@ -124,18 +67,18 @@ var/list/sector_shuttles = list()
+		if(possible_d.len)
+			D = input("Choose shuttle destination", "Shuttle Destination") as null|anything in possible_d
+		else
+			to_chat(usr,"<span class='warning'>No valid landing sites in range.</span>")
```
No return after this line or does the ```shuttle.set_destination_area(possible_d[D])``` call handle nulls?

comma replied:
Yeah, it has a nullcheck in beginning.

PsiOmegaDelta commented on code/modules/overmap/ships/computers/shuttle.dm:
```diff
@@ -124,18 +67,18 @@ var/list/sector_shuttles = list()
+		var/list/possible_d = shuttle.get_possible_destinations()
+		var/D
+		if(possible_d.len)
+			D = input("Choose shuttle destination", "Shuttle Destination") as null|anything in possible_d
```
It would be possible for a player to keep the popup open for a while and then come back ages later.
Should probably double check that D is still in the list of possible destinations as well as check ```CanInteract(user, default_state)```.

comma replied:
Yeah, gonna add that.

comma replied:
Checking if D is in list isn't needed since nulls are handled in set_destination, added interact check.

mwerezak replied:
There's a problem in that if the player leaves the dialog hanging and comes back ages later, the true list of possible destinations may now be different than the list stored in `possible_d` (because the ship has moved around in the meantime, for example). So `possible_d[D]` will be an invalid value, but a null check will not catch it.

So `set_destination_area(area/A)` probably needs to check that `A` is in fact still a valid possible destination.

comma committed:
Adds check to make sure we can still get the D

comma committed:
Generalizes shuttles a bit

Now you can define shuttles separately and set consoles to their tag in define.
If you don't, it defaults to current behavior.

comma committed:
Whoops

Bearcat is clearly superiour, but didn't mean to commit the map swap

comma committed:
Makes sure all shuttle areas are accessible

In case shuttle is defined separately, its starting area will be properly picked up by sector for landing_areas list

PsiOmegaDelta commented on code/modules/overmap/ships/computers/shuttle.dm:
```diff
@@ -124,18 +69,20 @@ var/list/sector_shuttles = list()
+		else
+			to_chat(usr,"<span class='warning'>No valid landing sites in range.</span>")
+		possible_d = shuttle.get_possible_destinations()
+		if(CanInteract(usr, default_state) && D in possible_d)
```
BYOND is finicky with that in operator, probably for the best to make this ```if(CanInteract(usr, default_state) && (D in possible_d))```.
Other than that the Topic issue appears resolved and the PR seems approvable.

comma replied:
Oh yeah, forgot to put them here, it always screws up if left unbracketed

comma committed:
Makes sure operations order is right
